dimod/cyutilities.pyx

```diff
@@ -0,0 +1,110 @@
+# Copyright 2021 D-Wave Systems Inc.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Un_lt required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+
+cimport cython
+
+import numpy as np
+
+
+cdef extern from "numpy/arrayobject.h":
+    ctypedef struct PyArray_Descr:
+        pass
+
+    object PyArray_Scalar(void*, PyArray_Descr*, object)
+
+np.import_array()  # needed for PyArray_Scalar
+
+# preconstruct these dtypes for speed, we could possibly improve it more
+# by casting to PyArray_Descr but I had trouble getting that to work
+cdef object _float32_dtype = np.dtype(np.float32)
+cdef object _float64_dtype = np.dtype(np.float64)
+
+cdef object as_numpy_float(cython.floating a):
+    if cython.floating == double:
+        return PyArray_Scalar(&a, <PyArray_Descr*>_float64_dtype, None)
+    elif cython.floating == float:
+        return PyArray_Scalar(&a, <PyArray_Descr*>_float32_dtype, None)
+    else:
+        raise NotImplementedError
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef inline bint _lt(Integer[:] row, Integer[:] col, Py_ssize_t i, Py_ssize_t j):
+    """Return (row[i], col[i]) < (row[j], col[j])."""
+    return row[i] < row[j] or (row[i] == row[j] and col[i] < col[j])
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef inline void _swap(Integer[:] row, Integer[:] col, cython.floating[:] data,
+                       Py_ssize_t i, Py_ssize_t j):
+    """Swap the values in the arrays at i, j"""
+    row[i], row[j] = row[j], row[i]
+    col[i], col[j] = col[j], col[i]
+    data[i], data[j] = data[j], data[i]
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cpdef Py_ssize_t coo_sort(Integer[:] row, Integer[:] col, cython.floating[:] data) except -1:
+    """Sort COO arrays in-place by row then by column."""
+    cdef Py_ssize_t num_interactions = row.shape[0]
+    if col.shape[0] != num_interactions or data.shape[0] != num_interactions:
+        raise ValueError("vectors should all be the same length")
+
+    # row index should be _lt than col index, this handles upper-triangular vs
+    # lower-triangular
+    cdef Py_ssize_t i
+    for i in range(num_interactions):
+        if row[i] > col[i]:
+            row[i], col[i] = col[i], row[i]
+
+    quicksort_coo(row, col, data, 0, num_interactions - 1)
+
+
+cdef void quicksort_coo(Integer[:] row, Integer[:] col, cython.floating[:] data,
```

Comment:
I do not know how important performance is, but it would be much better to use a non recursive function. The zip sort was not recursive. Although the performance boost will be less than 30%

Reply:
Yeah, this one is old. I know we could write a better performance one but it's never really been a bottleneck.